crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+}
+
+/// A stateful object to reach to an event, and then paginate backward and
+/// forward from it.
+///
+/// See also the module-level documentation.
+pub struct Paginator {
+    /// The room in which we're going to run the pagination.
+    room: Box<dyn PaginableRoom>,
+
+    /// Current state of the paginator.
+    state: SharedObservable<PaginatorState>,
+
+    /// The token to run the next backward pagination.
+    prev_batch_token: Mutex<Option<String>>,
+
+    /// The token to run the next forward pagination.
+    next_batch_token: Mutex<Option<String>>,
+}
+
+#[cfg(not(tarpaulin_include))]
+impl std::fmt::Debug for Paginator {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // Don't include the room in the debug output.
+        f.debug_struct("Paginator")
+            .field("state", &self.state.get())
+            .field("prev_batch_token", &self.prev_batch_token)
+            .field("next_batch_token", &self.next_batch_token)
+            .finish_non_exhaustive()
+    }
+}
+
+/// The result of a single pagination, be it from
+/// [`Paginator::paginate_backward`] or [`Paginator::paginate_forward`].
+#[derive(Debug)]
+pub struct PaginationResult {
+    /// Events returned during this pagination.
+    ///
+    /// If this is the result of a backward pagination, then the events are in
+    /// reverse topological order.
+    ///
+    /// If this is the result of a forward pagination, then the events are in
+    /// topological order.
+    pub events: Vec<TimelineEvent>,
+
+    /// Did we hit an end of the timeline?
+    ///
+    /// If this is the result of a backward pagination, this means we hit the
+    /// *start* of the timeline.
+    ///
+    /// If this is the result of a forward pagination, this means we hit the
+    /// *end* of the timeline.
+    pub hit_end_of_timeline: bool,
+}
+
+/// The result of an initial [`Paginator::start_from`] query.
+#[derive(Debug)]
+pub struct StartFromResult {
+    /// All the events returned during this pagination, in topological ordering.
+    pub events: Vec<TimelineEvent>,
+
+    /// Whether the /context query returned a previous batch token.
+    pub has_prev: bool,
+
+    /// Whether the /context query returned a next batch token.
+    pub has_next: bool,
+}
+
+impl Paginator {
+    /// Create a new [`Paginator`], given a room implementation.
+    pub fn new(room: Box<dyn PaginableRoom>) -> Self {
+        Self {
+            room,
+            state: SharedObservable::new(PaginatorState::Initial),
+            prev_batch_token: Mutex::new(None),
+            next_batch_token: Mutex::new(None),
+        }
+    }
+
+    /// Check if the current state of the paginator matches the expected one.
+    fn check_state(&self, expected: PaginatorState) -> Result<(), PaginatorError> {
+        let actual = self.state.get();
+        if actual != expected {
+            return Err(PaginatorError::InvalidPreviousState { expected, actual });
```

Comment:
Personal preference, but I'd prefer an `else` on this `if` to avoid the `return`.